### PR TITLE
[Draft] Second version of cupy support for sparse matrix cg sampling

### PR DESCRIPTION
**Pull request by chinandrew**
Unlike #15, this PR adds the use_cupy flag to the model/matrix, and then has the dot function handle the types of the input vectors case by case. This way cg_sampler can just use the design matrix flags without needing a use_cupy flag. Theres' probably cleaner way to do the logic, but as a prototype it seems to work and is actually faster than #15.

### CODE REVIEW

**Comment by aki-nishimura** [comment]
I refactored the code a bit, but otherwise the cupy integration looks good to me. Perhaps you can go ahead and ff-merge after squashing commits to clean things up? It seems reasonable to just squash all the commits since `master` into one unless you see smaller coherent units.

Incidentally, an idea for even cleaner approach going forward. It seems like we have managed to completely get rid of the scipy matrix when using cupy? In this case, instead of having the `use_cupy` flag, maybe the `RegressionModel()` should just accept cupy matrix as an input and the whole module should adjust its behavior based on the type of `X`.  The `self.use_cupy` could even be a property that checks if `X` is a cupy array or not.

**Comment by chinandrew** [comment]
> I refactored the code a bit, but otherwise the cupy integration looks good to me. Perhaps you can go ahead and ff-merge after squashing commits to clean things up? It seems reasonable to just squash all the commits since `master` into one unless you see smaller coherent units.


Yea squashing works for me when it's ready.

> Incidentally, an idea for even cleaner approach going forward. It seems like we have managed to completely get rid of the scipy matrix when using cupy? In this case, instead of having the `use_cupy` flag, maybe the `RegressionModel()` should just accept cupy matrix as an input and the whole module should adjust its behavior based on the type of `X`. The `self.use_cupy` could even be a property that checks if `X` is a cupy array or not.


I just tried giving this a shot by adding a `is_cupy_sparse` method to abstract matrix (wasn't quite sure where the best place to put it was, open to suggestions) that gets called by the sparse design matrix and also by the factory. Also updated some various logic to handle cupy input matrices (e.g. removing the intercept). Let me know what you think, we can always revert back to the previous commit if we think it needs more work.

File: bayesbridge/design_matrix/abstract_matrix.py [hunk 1]
```diff
@@ -63,12 +70,16 @@ def toarray(self):
...
+    def is_cupy_sparse(X):
+        return isinstance(X, cpx.scipy.sparse.spmatrix) if cp and cpx else False
+
     @staticmethod
     def remove_intercept_indicator(X):
-        if sp.sparse.issparse(X):
-            col_variance = np.squeeze(np.array(
-                X.power(2).mean(axis=0) - np.power(X.mean(axis=0), 2)
-            ))
+        squeeze, array, power = (cp.squeeze, cp.array, cp.power) if \
```
Comment (chinandrew) [inline bayesbridge/design_matrix/abstract_matrix.py:79]: at some point it might be easier to doing something like `linalg_pkg = cp if self.use_cupy else np`...

**Comment by chinandrew** [comment]
Did some minor updates to logics/checks around use_cupy to catch edge cases and clean up messages, added some basic gpu-dependent tests (pass on both my machine and jhpce),  and added cupy as an extra optional dependency to setup.py

File: bayesbridge/design_matrix/sparse_matrix.py [hunk 2]
```diff
@@ -73,7 +68,7 @@ def dot(self, v):
                 return self.X_dot_v
             self.v_prev = v.copy()
 
-        input_is_cupy = (cp is not None) and isinstance(v, cp._core.core.ndarray)
+        input_is_cupy = (cp is not None) and isinstance(v, cp.ndarray)
```
Comment (chinandrew) [inline bayesbridge/design_matrix/sparse_matrix.py:71]: turns out this works

File: bayesbridge/design_matrix/abstract_matrix.py [hunk 3]
```diff
@@ -63,12 +70,16 @@ def toarray(self):
...
+            AbstractDesignMatrix.is_cupy_sparse(X) else (np.squeeze, np.array, np.power)
+        if sp.sparse.issparse(X) or AbstractDesignMatrix.is_cupy_sparse(X):
+            col_variance = squeeze(array(X.power(2).mean(axis=0) - power(X.mean(axis=0), 2)))
```
Comment (aki-nishimura) [inline bayesbridge/design_matrix/abstract_matrix.py:91]: Would the following work (w/o `{np, cp}.power()`)?
```suggestion
            col_variance = squeeze(array(X.power(2).mean(axis=0) - X.mean(axis=0).power(2)))
```

File: bayesbridge/design_matrix/abstract_matrix.py (hunk 1 above)
Comment (aki-nishimura) [reply bayesbridge/design_matrix/abstract_matrix.py:88]: `linalg_pkg = cp if self.use_cupy else np` is a great idea. (The name `linalg_pkg` might be confusing, though; it sounds like a type of BLAS. Off top of my head, `array_lib` comes to my mind, though still not great.)

File: tests/gpu_tests/test_gibbs.py [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
+except (ImportError, ModuleNotFoundError) as e:
+    cp = None
+    cupy_exception = e
 import numpy as np
 import pytest
 
 from ..helper import simulate_data
 
 
+pytestmark = pytest.mark.skipif(cp is None, reason="cupy unavailable")
```
[the remark this answers is not in the document]
Comment (chinandrew) [reply tests/gpu_tests/test_gibbs.py:12]: Good call to use pytest skip functionality. Probably easiest to just put `pytest.importorskip("cupy")` at the top of the file, just pushed a commit for it.

File: bayesbridge/design_matrix/abstract_matrix.py (hunk 3 above)
Comment (chinandrew) [reply bayesbridge/design_matrix/abstract_matrix.py:91]: I don't think so, `power()` doesnt appear to be a method for matrices (for np and cp).
```
In [10]: X = np.ones((5,5))

In [11]: X.mean(axis=0).power(2)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-2fff9e220fa4> in <module>
----> 1 X.mean(axis=0).power(2)

AttributeError: 'numpy.ndarray' object has no attribute 'power'
```

Comment (aki-nishimura) [reply bayesbridge/design_matrix/abstract_matrix.py:91]: Ah, silly me. Thanks, good to know.
